compiler/src/iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPUSelectUKernels.cpp

```diff
@@ -0,0 +1,155 @@
+// Copyright 2024 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Codegen/LLVMGPU/Utils/LLVMGPUSelectUKernels.h"
+
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h"
+#include "iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenDialect.h"
+#include "iree/compiler/Codegen/Utils/GPUUtils.h"
+#include "iree/compiler/Codegen/Utils/Utils.h"
+#include "iree/compiler/Utils/EmbeddedDataDirectory.h"
+#include "llvm/Support/FormatVariadic.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/IR/AsmState.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinAttributes.h"
+
+namespace mlir::iree_compiler {
+
+namespace {
+
+static StringLiteral executableObjectsAttrName = "hal.executable.objects";
```

Comment:
```suggestion
constexpr StringLiteral executableObjectsAttrName = "hal.executable.objects";
```

constexpr in global scope already implies static

Reply:
It should be possible to design a `StringLiteral` class that doesn't have this problem. It's just that it needs to store the `const char[N]` by value, and be templated in N, instead of storing a (inheriting from) `StringRef`.

Reply:
I thought that `static constexpr` always meant no dynamic initialization?

Reply:
This is very much the intended usecase for StringLiteral -- constant names that eventually get passed as StringRef (without the need of ever calling strlen).

Reply:
Whenever address-of is "constexpr", there is, one way or another, a hidden caveat. It simply isn't possible to tell the address at compile time. What can be known at compile time is the offset at which something is allocated relatively to the start of a segment, but the segment base address won't be know until the DSO is loaded. So at runtime, once that is known, the base+offset addition can be computed, and its result written as the actual initializer for such pointer constants. IIUC that is what toolchain experts call a "relocation", but I am talking out of my depth here, so please feel free to point out any misunderstanding.

Reply:
I tried all possible combinations of storage class and type for this `executableObjectsAttrName`,
```
{static, constexpr, none} x {StringLiteral, StringRef, const char[], const char*}
```
and that makes no difference to the number of relocations in `libIREECompiler.so`.